vyper/semantics/validation/annotation.py

```diff
@@ -96,6 +97,23 @@ def visit_For(self, node):
             self.expr_visitor.visit(node.iter)
 
 
+def validate_address_code_attribute(node: vy_ast.Attribute, type_: BaseTypeDefinition) -> None:
```

Comment:
I think this functionality is more appropriate in `vyper/semantics/validation/local.py`. It would be very clean if this functionality were put in the following function: https://github.com/vyperlang/vyper/blob/2f1584bf9d2fce15e2173b8befc83619c5deae79/vyper/semantics/validation/local.py#L182-L191.

Reply:
Addressed in https://github.com/vyperlang/vyper/pull/2583/commits/5a5798d34b09587bea05af7e71240c1c1d554a36

Reply:
It seems I have a bug whichever `ExpressionAnnotationVisitor` or `FunctionNodeVisitor` I choose.
With `ExpressionAnnotationVisitor`, I think some attribute nodes won't get traversed and miss the validation.
With `FunctionNodeVisitor`, as showed up on CI, `get_exact_type_from_node` will fail on the second `visit` since it doesn't setup correct namespaces.

Reply:
I don't think any typechecking should be done in the annotation visitor since that merely propagates types to nodes. I'm a bit confused why you're code wouldn't work in `FunctionNodeVisitor`. Is there some bug regarding typechecking in loops?

Reply:
I haven't looked into it but I was just in the process of fixing some typechecking for list literals. cf. https://github.com/vyperlang/vyper/pull/2584#issuecomment-1003938352, https://github.com/vyperlang/vyper/pull/2584#issuecomment-1004528998 and https://github.com/vyperlang/vyper/issues/2587

Reply:
@hi-ogawa could you merge in master now that #2556 is merged and see if that fixes the issue

Reply:
Thanks for letting me know about that! I can get back to this tomorrow and I will check that out if it works. Sorry for the delay.
Regarding the reason why `FunctionNodeVisitor` doesn't work, I think that's because of the structure around this:

```py
 def visit(self, node): 
     super().visit(node)   #  <<<=== 1st visit (possibly `Attribute` won't be visited this time)
     self.annotation_visitor.visit(node) 
  
     attr_descendants = node.get_descendants(vy_ast.Attribute) 
     for attr_descendant in attr_descendants: 
         self.visit(attr_descendant)    # <<<=== 2nd visit (this is to make sure to visit at least all the `Attribute` nodes missed on 1st visit above)
```

The way visiting the `Attribute` on 2nd time doesn't setup namespace properly. For example, from CI,

https://github.com/vyperlang/vyper/runs/4707801642?check_suite_focus=true#step:5:6264

```py
for i in ...:
  ... _baz[i].x ...  # <<<=== try to visit this `Attribute` node
```

here, this tries to visit `Attribute` node `_baz[i].x` without setting up the namespace for the for-loop variable `i` and thus `get_exact_type_from_node` on the value node `_baz[i]` is failing.

I haven't actually debug it, so I might be wrong, but this is my intuition of the issue.

Reply:
@charles-cooper Sorry for the break and thanks the patience.
I pushed the commit https://github.com/vyperlang/vyper/pull/2583/commits/f13e4f78dc55c5e3ae1841f55f36fd36ac77adb5 for the fix. I modified `FunctionNodeVisitor` (with `FunctionNodeExpressionVisitor`) to visit all the nodes up in the single pass.
I found this more natural approach than grabbing `vy_ast.Attribute` descendent nodes and validating certain properties.
I might be missing some reasons why it hasn't implemented in this way, so I would appreciate if you could review this change for potential drawbacks.
Thanks a lot!

Reply:
I think your approach is natural and makes sense. It probably wasn't implemented that way because we didn't really have context-dependent checks like that. The only thing I would ask is to change the name of the class to indicate that it is private to this module - maybe `_LocalExprVisitor`